### PR TITLE
Content Blocker improvements

### PR DESCRIPTION
**Pull request by bwaresiak**
Task/Issue URL: https://app.asana.com/0/414235014887631/1200341959926298
Tech Design URL: https://app.asana.com/0/414235014887631/1200361182088351
CC: 

**Description**:
Rewrite of Content Blocker Manager and supporting files to handle edge cases, improve thread safety and performance. See tech design for details.

**Steps to test this PR**:

Generic smoke tests.

Ensure tests are handling all expected and edge cases.

Look into:
- Content blocker effectiveness .
- Performance when cold starting.
- Performance when updating lists/toggling protection.
- Handling TDS and TempList updates (error cases, lost connections etc).

Test onboarding, to ensure https://app.asana.com/0/414235014887631/1200195915061851 glitch has been fixed.

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [ ] iPad

**OS Testing**:

* [x] iOS 13
* [x] iOS 14

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
**When ready for review, remember to post the PR in MM**

### CODE REVIEW

File: Core/ContentBlockerRulesManager.swift [hunk 1]
```diff
@@ -23,105 +23,318 @@ import os.log
...
+        workQueue.async {
+            self.requestCompilation()
         }
     }
 
-    /// Return compiled rules for the current content blocking configuration.  This may return a precompiled rule set.
-    public func compiledRules(completion: ((WKContentRuleList?) -> Void)?) {
-        guard !isCompilingRules else {
-            completion?(nil)
+    private func requestCompilation() {
```
Comment (brindy) [inline Core/ContentBlockerRulesManager.swift:89]: If the state is already `.recompilingAndScheduled` could we just skip this?

File: Core/ContentBlockerRulesManager.swift [hunk 2]
```diff
@@ -23,105 +23,318 @@ import os.log
...
+                for id in idsSet {
+                    WKContentRuleListStore.default()?.removeContentRuleList(forIdentifier: id) { _ in }
+                }
+            })
+        }
     }
 
 }
+
+extension ContentBlockerRulesManager {
```
Comment (brindy) [inline Core/ContentBlockerRulesManager.swift:315]: This extension appears to be just for tests ... is it intended to live here?

**Review by brindy — COMMENTED** [review]
Looking good so far, going to run it this evening to monitor it.  Just a couple of minor questions for you.

**Review by brindy — COMMENTED** [review]
We talked about some stuff over zoom and I've tested a clean install on my iPad.  Looks good to me.  Will let @samsymons give the final 👍

File: Core/ContentBlockerRulesManager.swift (hunk 1 above)
Comment (bwaresiak) [reply Core/ContentBlockerRulesManager.swift:89]: We do, it is in the guard :)

File: Core/ContentBlockerRulesManager.swift (hunk 2 above)
Comment (bwaresiak) [reply Core/ContentBlockerRulesManager.swift:315]: We agreed to keep it, as it references fileprivate API that would otherwise have to be internal.

**Review by samsymons — APPROVED** [review]
LGTM, tested this a fair bit locally without issue. Nice work, this is a well thought out implementation. 👍

**Comment by bwaresiak** [comment]
Thank you, @samsymons @brindy
